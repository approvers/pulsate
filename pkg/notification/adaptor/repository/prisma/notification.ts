import { Ether, Option, Result } from '@mikuroxina/mini-fn';
import type { Prisma, PrismaClient } from '@prisma/client';
import type { AccountID } from '../../../../accounts/model/account.js';
import {
  Notification,
  type NotificationActorType,
  type NotificationID,
  type NotificationType,
  type NotificationTypeMap,
} from '../../../model/notification.js';
import {
  NOTIFICATION_DEFAULT_LIMIT,
  NOTIFICATION_MAX_LIMIT,
  type NotificationFilter,
  type NotificationRepository,
  notificationRepoSymbol,
} from '../../../model/repository/notification.js';

/**
 * NOTE:
 * - 0 is reserved.
 * - followed -> 1
 * - followRequested -> 2
 * - followAccepted -> 3
 * - mentioned -> 4
 * - renoted ->  5
 * - reacted ->  6
 */
const NOTIFICATION_TYPE_MAP = {
  followed: 1,
  followRequested: 2,
  followAccepted: 3,
  mentioned: 4,
  renoted: 5,
  reacted: 6,
} as const satisfies Record<NotificationType, number>;
const NOTIFICATION_TYPE_CODE_MAP: Record<number, NotificationType> = [];

for (const [k, v] of Object.entries(NOTIFICATION_TYPE_MAP)) {
  if (v in NOTIFICATION_TYPE_CODE_MAP) throw new Error('Duplicate value');

  NOTIFICATION_TYPE_CODE_MAP[v] = k as NotificationType;
}
Object.freeze(NOTIFICATION_TYPE_CODE_MAP);

export class PrismaNotificationRepository implements NotificationRepository {
  private readonly prisma: PrismaClient;

  constructor(prisma: PrismaClient) {
    this.prisma = prisma;
  }

  private serialize(
    notification: Notification,
  ): Prisma.NotificationCreateInput {
    return {
      id: notification.getID(),
      recipient: {
        connect: {
          id: notification.getRecipientID(),
        },
      },
      notificationType:
        NOTIFICATION_TYPE_MAP[notification.getNotificationType()],
      actorType: notification.getActorType(),
      actor: {
        connect: {
          id: notification.getActorID(),
        },
      },
      sourceID: notification.getSourceID() ?? '',
      activityID: notification.getActivityID() ?? '',
      createdAt: notification.getCreatedAt(),
      readAt: null,
    };
  }

  private serializeFilter(
    filter: NotificationFilter,
  ): Prisma.NotificationFindManyArgs {
    const orderBy = () => {
      if (Option.isNone(filter.cursor)) {
        // default: desc
        return 'desc';
      }
      const cursor = Option.unwrap(filter.cursor);
      if (cursor.type === 'before') {
        return 'desc';
      }

      return 'asc';
    };
    const cursor = () => {
      if (Option.isNone(filter.cursor)) {
        return undefined;
      }
      const cursor = Option.unwrap(filter.cursor);
      if (cursor.type === 'before') {
        return {
          cursor: {
            id: cursor.id,
          },
          skip: 1,
        };
      }

      return {
        cursor: {
          id: cursor.id,
        },
        skip: 1,
      };
    };

    return {
      orderBy: {
        // default: desc, if afterID specified: asc
        createdAt: orderBy(),
      },
      ...cursor(),
      take: Option.isSome(filter.limit)
        ? Option.unwrap(filter.limit)
        : NOTIFICATION_DEFAULT_LIMIT,
    };
  }

  private deserialize(
    notification: Prisma.PromiseReturnType<
      typeof this.prisma.notification.findUnique
    >,
  ): Result.Result<Error, Notification> {
    if (!notification) {
      // ToDo: Define NotificationNotFoundError
      return Result.err(new Error('Notification not found'));
    }
    const notificationType =
      NOTIFICATION_TYPE_CODE_MAP[notification.notificationType];
    if (!notificationType) {
      return Result.err(new Error('Invalid notification type'));
    }

    return Result.ok(
      Notification.reconstruct({
        id: notification.id as NotificationID,
        recipientID: notification.recipientID as AccountID,
        notificationType,
        actorType: notification.actorType as NotificationActorType,
        actorID: notification.actorID as AccountID,
        createdAt: notification.createdAt,
        readAt: notification.readAt
          ? Option.some(notification.readAt)
          : Option.none(),
        sourceID:
          notification.sourceID === ''
            ? null
            : (notification.sourceID as NotificationTypeMap[typeof notificationType]['source']),
        activityID:
          notification.activityID === ''
            ? null
            : (notification.activityID as NotificationTypeMap[typeof notificationType]['activity']),
      }),
    );
  }

  async create(
    notification: Notification,
  ): Promise<Result.Result<Error, void>> {
    try {
      await this.prisma.notification.create({
        data: this.serialize(notification),
      });
      return Result.ok(undefined);
    } catch (e) {
      return Result.err(e as unknown as Error);
    }
  }

  async findByID(
    id: NotificationID,
  ): Promise<Result.Result<Error, Notification>> {
    try {
      const res = await this.prisma.notification.findUnique({
        where: {
          id: id,
        },
      });
      return this.deserialize(res);
    } catch (e) {
      return Result.err(e as unknown as Error);
    }
  }
  async findByRecipientID(
    recipientID: string,
    filter: NotificationFilter,
  ): Promise<Result.Result<Error, Notification[]>> {
    try {
      if (Option.isSome(filter.limit)) {
        if (Option.unwrap(filter.limit) > NOTIFICATION_MAX_LIMIT) {
          // ToDo: Define NotificationCursorLimitOutOfRangeError
          return Result.err(new Error('Limit exceeds the maximum value'));
        }
      }

      const res = await this.prisma.notification.findMany({
        where: {
          recipientID: recipientID,
        },
        ...this.serializeFilter(filter),
      });

      const notificationsRes = res.map((n) => this.deserialize(n));
      const notifications = notificationsRes
        .filter(Result.isOk)
        .map(Result.unwrap);

      if (notificationsRes.length !== notifications.length) {
        return Result.err(new Error('Failed to deserialize'));
      }

      return Result.ok(notifications);
    } catch (e) {
      return Result.err(e as unknown as Error);
    }
  }

  async updateReadAt(
    notification: Notification,
  ): Promise<Result.Result<Error, void>> {
    try {
      if (Option.isNone(notification.getReadAt())) {
        // ToDo: Define NotificationReadAtNotSetError
        return Result.err(new Error('ReadAt is not set'));
      }

      await this.prisma.notification.update({
        where: {
          id: notification.getID(),
        },
        data: {
          readAt: Option.unwrap(notification.getReadAt()),
        },
      });
      return Result.ok(undefined);
    } catch (e) {
      return Result.err(e as unknown as Error);
    }
  }
}
export const prismaNotificationRepo = (client: PrismaClient) =>
  Ether.newEther(
    notificationRepoSymbol,
    () => new PrismaNotificationRepository(client),
  );
