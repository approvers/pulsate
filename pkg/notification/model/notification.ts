import { Option } from '@mikuroxina/mini-fn';
import type { AccountID } from '../../accounts/model/account.js';
import type { ID } from '../../id/type.js';
import type { NoteID } from '../../notes/model/note.js';
import type { ReactionID } from '../../notes/model/reaction.js';

export type NotificationID = ID<Notification>;
export type NotificationType =
  | 'followed'
  | 'followRequested'
  | 'followAccepted'
  | 'mentioned'
  | 'renoted'
  | 'reacted';

export type NotificationTypeMapValue<Source, Activity> = {
  source: Source;
  activity: Activity;
};

export type NotificationTypeMap = {
  followed: NotificationTypeMapValue<null, null>;
  followRequested: NotificationTypeMapValue<null, null>;
  followAccepted: NotificationTypeMapValue<null, null>;
  mentioned: NotificationTypeMapValue<null, NoteID>;
  renoted: NotificationTypeMapValue<NoteID, NoteID>;
  reacted: NotificationTypeMapValue<NoteID, ReactionID>;
  unknown: NotificationTypeMapValue<never, never>;
};

export type NotificationActorType = 'account' | 'system';

export interface CreateNotificationArgs<T extends NotificationType> {
  /**
   * Notification ID
   */
  id: NotificationID;
  /**
   * Recipient account ID
   */
  recipientID: AccountID;
  /**
   * Notification Type
   * @description
   * - followed: Followed
   * - followRequested: Follow requested
   * - followAccepted: Follow accepted
   * - mentioned: Mentioned
   * - renoted: Renoted
   * - reacted: Reacted
   */
  notificationType: T;
  /**
   *
   */
  sourceID: NotificationTypeMap[T]['source'];
  activityID: NotificationTypeMap[T]['activity'];

  /**
   * Actor(who did the notified action) Type
   * @description
   * - account: Account
   * - system: System
   */
  actorType: NotificationActorType;
  /**
   * Actor account ID
   */
  actorID: AccountID;
  /**
   * Created At
   */
  createdAt: Date;
  /**
   * Read At
   */
  readAt: Option.Option<Date>;
}

export class Notification {
  private readonly id: NotificationID;
  private readonly notificationType: NotificationType;
  private readonly recipientID: AccountID;
  private readonly createdAt: Date;
  private readonly actorType: NotificationActorType;
  private readonly actorID: AccountID;

  private readonly sourceID: NotificationTypeMap[typeof this.notificationType]['source'];
  private readonly activityID: NotificationTypeMap[typeof this.notificationType]['activity'];

  private readAt: Option.Option<Date>;

  private constructor(args: CreateNotificationArgs<NotificationType>) {
    this.id = args.id;
    this.recipientID = args.recipientID;
    this.notificationType = args.notificationType;
    this.actorType = args.actorType;
    this.actorID = args.actorID;
    this.createdAt = args.createdAt;
    this.readAt = args.readAt;

    this.sourceID = args.sourceID;
    this.activityID = args.activityID;
  }

  static new<T extends NotificationType>(
    args: Omit<CreateNotificationArgs<T>, 'isRead' | 'readAt'>,
  ) {
    return new Notification({
      ...args,
      readAt: Option.none(),
    });
  }

  static reconstruct<T extends NotificationType>(
    args: CreateNotificationArgs<T>,
  ) {
    return new Notification(args);
  }

  /**
   * Get Notification ID
   * @example
   * ```
   * "349875930483"
   * ```
   */
  getID(): NotificationID {
    return this.id;
  }

  /**
   * Get recipient account ID
   */
  getRecipientID(): AccountID {
    return this.recipientID;
  }

  /**
   * Get Notification Type
   * @description
   * - followed: Followed
   * - followRequested: Follow requested
   * - followAccepted: Follow accepted
   * - mentioned: Mentioned
   * - renoted: Renoted
   * - reacted: Reacted
   */
  getNotificationType(): NotificationType {
    return this.notificationType;
  }

  /**
   * Get Source ID
   * @description
   * SourceID is the target of the Activity
   *
   * e.g.
   * - notification type is "followed", SourceID is null (Source is Account, but it's already known by recipient)
   * - notification type is "reacted", SourceID is NoteID
   */
  getSourceID(): NotificationTypeMap[typeof this.notificationType]['source'] {
    return this.sourceID;
  }

  /**
   * Get Activity ID
   * @description
   * ActivityID is the object resulting from the operation on Source.
   *
   * e.g.
   * - notification type  "followed", ActivityID is null (no objects are generated when followed)
   * - notification type  "reacted", ActivityID is ReactionID (Reaction is the object generated by the operation)
   */
  getActivityID(): NotificationTypeMap[typeof this.notificationType]['activity'] {
    return this.activityID;
  }

  getCreatedAt(): Date {
    return this.createdAt;
  }

  /**
   * Get Actor(who did the notified action) Type
   * @description
   * - account: Account
   * - system: System
   */
  getActorType(): NotificationActorType {
    return this.actorType;
  }

  /**
   * Get Actor account ID
   */
  getActorID(): AccountID {
    return this.actorID;
  }

  /**
   * Get Read Status
   */
  getIsRead(): boolean {
    return Option.isSome(this.readAt);
  }

  /**
   * Set time when message read
   *
   * NOTE: Once a message has been read, it cannot be marked as unread.
   */
  setRead(date: Date) {
    this.readAt = Option.some(date);
  }

  /**
   * Get date when the notification was read
   */
  getReadAt(): Option.Option<Date> {
    return this.readAt;
  }
}
