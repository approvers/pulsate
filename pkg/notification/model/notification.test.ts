import { describe, expect, it } from 'vitest';
import type { AccountID } from '../../accounts/model/account.js';
import { Notification, type NotificationID } from './notification.js';

describe('Notification', () => {
  it('should create new instance', () => {
    const res = Notification.new({
      id: '1' as NotificationID,
      notificationType: 'followed',
      recipientID: '10' as AccountID,
      createdAt: new Date('2023-09-10T00:00:00.000Z'),
      actorType: 'account',
      actorID: '20' as AccountID,
      sourceID: null,
      activityID: null,
    });
    expect(res).toMatchSnapshot();
  });
});
