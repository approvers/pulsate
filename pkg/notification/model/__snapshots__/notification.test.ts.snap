// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Notification > should create new instance 1`] = `
Notification {
  "activityID": null,
  "actorID": "20",
  "actorType": "account",
  "createdAt": 2023-09-10T00:00:00.000Z,
  "id": "1",
  "notificationType": "followed",
  "readAt": [
    Symbol(OptionNone),
  ],
  "recipientID": "10",
  "sourceID": null,
}
`;
